platform/posix: switch from getchar() to linenoise based console

SalimTerryLi opened the pull request:
### Changes

- bring the modified source of linenoise in-tree
- server_io.cpp added new get_stdin() method, which is primarily used by uorb top. Now uorb top can work with MavShell inputs
- apps.cpp.in use PX4_INFO_RAW() to ensure all outputs go into thread stdout, which will be redirected to its real location

### Improvements

- MavlinkShell won't dup2() on original stdio, thus will not mess the real terminal
- use of the linenoise makes it much easier to implement advanced terminal functionaliies

Now the console can be gracefully shutdown, but further cleanups still can not be performed until per-thread graceful shutdown is implemented.

### Test coverage

- works on native arm64 RPi 4B
- `make px4_sitl gz_x500` works on x64 with Linux Mint 21
- `make px4_sitl jmavsim` and `make px4_sitl gz_x500`works on Intel MacBookPro 2017 with macOS 12.6.5

SalimTerryLi commented:
Have no idea how to fix CI and let it pull my submodule....

SalimTerryLi commented:
Have no idea how to fix CI and let it pull my submodule....

dagar commented:
TODO:
 - verify licensing implications
 - bring code in tree or fork submodule under PX4 org (as a precaution)

SalimTerryLi commented:
> TODO:
> 
>     * verify licensing implications
> 
>     * bring code in tree or fork submodule under PX4 org (as a precaution)

According to the following resources:

- https://www.gnu.org/licenses/license-compatibility.en.html
- https://www.gnu.org/licenses/license-list.en.html#ModifiedBSD
- https://www.gnu.org/licenses/license-list.en.html#FreeBSD
- https://en.wikipedia.org/wiki/BSD_licenses

BSD2 should be compatible with, and even more flexible than BSD3, so that it should be fine to include the linenoise project into PX4.

A live example is here: https://github.com/espressif/esp-idf/tree/master/components/console/linenoise

Espressif includes this project into its Apache2 licensed product.

About the second question: Whether in-tree or as-a-module?

I'd prefer to leave it as an independent module, as I've fixed a lot to make that old project works with multi-instances and brings pipe support to it. Now it has a different API, comparing to the original project. If it is possible, I'd want it to work outside PX4 as well.

If possible, how could this project be forked into PX4 org?

SalimTerryLi commented:
Seems the only left issue here is clang-tidy. Is it possible to bypass the check for external submodules? Or should it conformant to the PX4 style as well? (This case it's better to move the library in-tree instead)

SalimTerryLi committed:
platform/posix: switch from getchar() to linenoise based console

Changes:

- bring the midified source of linenoise in-tree, with clang-tidy
  and astyle check disabled
- server_io.cpp added new get_stdin() method, which is primarily
  used by uorb top. Now uorb top can work with MavShell inputs
- apps.cpp.in use PX4_INFO_RAW() to ensure all outputs go into
  thread stdout, which will be redirected to its real location

Improvements:

- MavlinkShell won't dup2() on original stdio, thus will not mess
  the real terminal
- use of the linenoise makes it much easier to implement advanced
  terminal functionaliies

Limitations:

- Currently stderr is not managed by console as there is no real
  usage within PX4.

Now the console can be gracefully shutdown, but further cleanups
still can not be performed until per-thread graceful shutdown
is implemented.
